apis/python/src/tiledbsoma/_dask/util.py

```diff
@@ -0,0 +1,147 @@
+# Copyright (c) TileDB, Inc. and The Chan Zuckerberg Initiative Foundation
+#
+# Licensed under the MIT License.
+
+"""Utilities for loading a TileDB-SOMA ``X`` matrix as a Dask Array.
+"""
+
+from __future__ import annotations
+
+from concurrent.futures import ThreadPoolExecutor
+from dataclasses import dataclass, field
+from functools import cache
+from os import cpu_count
+from typing import (
+    Any,
+    Sequence,
+    Union,
+)
+
+import numpy as np
+import pyarrow as pa
+from numpy import int64
+from numpy.typing import NDArray
+from somacore.options import SparseNDCoord, SparseNDCoords
+from typing_extensions import TypeAlias
+
+from tiledbsoma.options import SOMATileDBContext
+from tiledbsoma.options._soma_tiledb_context import ConfigVal
+
+ChunkSize = Union[int, tuple[Union[int, None], int]]
+JoinIDs: TypeAlias = NDArray[int64]
+
+
+@dataclass
+class DaskConfig:
+    chunk_size: ChunkSize
+    tdb_concurrency: int | None = 1
+    tdb_configs: dict[str, ConfigVal] = field(default_factory=dict)
+
+
+def chunk_ids_sizes(
+    joinids: JoinIDs, chunk_size: int, dim_size: int
+) -> tuple[list[JoinIDs], list[int]]:
+    """Slice chunks from joinids, return chunks' joinids and sizes."""
+    chunk_joinids: list[JoinIDs] = []
+    chunk_sizes: list[int] = []
+    i = 0
+    n = len(joinids)
+    while i < n:
+        end_idx = min(n, i + chunk_size)
+        num = end_idx - i
+        chunk = joinids[i:end_idx]
+        chunk_joinids.append(chunk)
+        chunk_sizes.append(num)
+        i += num
+    return chunk_joinids, chunk_sizes
+
+
+def make_context(
+    tdb_concurrency: int | None,
+    tiledb_config: dict[str, Any],
+) -> SOMATileDBContext:
+    """Create and cache a ``SOMATileDBContext``, optionally setting several TileDB concurrency configs.
+
+    Intended for use with ``load_daskarray``, where we want to restrict TileDB concurrency when running TileDB-SOMA
+    queries from many Dask workers in parallel. In such cases we don't want each TileDB-SOMA query thinking it should
+    utilize all ``cpu_count`` CPUs.
+
+    This wrapper just flattens the ``tiledb_config`` ``dict`` into hashable ``tuple``s (for use with ``cache``).
+    """
+    return _make_context(
+        tdb_concurrency=tdb_concurrency,
+        tiledb_configs=tuple(tiledb_config.items()),
+    )
+
+
+@cache
+def _make_context(
+    tdb_concurrency: int | None,
```

Comment:
Also here (and throughout) I'd prefer we replace `tdb_` with `tiledb_`. True throughout this PR; here, for an additional reason which is that one of these uses `tdb_` and the other uses `tiledb_`, and consistency would be nice.